---
title: "SLC35A2 Olig2 cKO snRNAseq analysis"
output: html_notebook
editor_options: 
  chunk_output_type: console
---
# Project set up

## Random Seed
```{r Setting seed, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}}
set.seed(0827)
```
## Library load
```{r Library loading, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
#### Library load ####
# Analysis
library(parallel)
library(future)
library(Seurat)
library(glmGamPoi)
library(SeuratExtend)
library(clusterProfiler)
library(DoubletFinder)
library(PCAtools)

# library(rrvgo)
library(SoupX)
library(miloR)
library(SingleCellExperiment)
library(scater)
library(scran)

# Stat test
library(presto)
library(speckle)
library(scProportionTest)

# Reference
library(org.Mm.eg.db)
library(BSgenome.Mmusculus.UCSC.mm10)

# File load and save
library(qs)

# Load HDF5 library from IGM
# dyn.load("/igm/apps/hdf5/hdf5-1.12.1/lib/libhdf5_hl.so.200")  
library(hdf5r)

# Data manipulation
library(tidyverse)

# Visualization
library(ggplot2)
library(ggstats)
library(patchwork)
library(EnhancedVolcano)
library(enrichplot)
library(ggtext)
library(cowplot)
library(ggpubr)
library(reshape2)
```

```{r}
options(Ncpus = parallel::detectCores()-1)
options(future.globals.maxSize = 16000 * 1024^2)
```

## Source Code list
```{r Source Code list, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Load custom function scripts
source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/01__load_rna_from_h5.R")
source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/02__create_seurat_object.R")
source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/03__rna_qc.R")
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/04__process_rna.R')
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/05__rpca_rna_integration.R')
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/06__process_integrated.R')
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/07__clustering.R')
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/08__ref_mousebrain_annotation.R')
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/99_doublet_process_rna.R')

```

## Folder directory
```{r directory list, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Directory setup
h5_input_dir <- 
  '/igm/projects/250718_GSL-EM-4483'
save_dir <- 
  '~/SLC35A2_Olig2cKO_snRNA_JY/WIP'
qsave_dir <- 
  file.path(save_dir, "QSAVE")
qc_dir    <- 
  file.path(save_dir, "QC")
csv_dir <- 
  file.path(save_dir, 'CSV')
plot_dir <- 
  file.path(save_dir, 'PLOTS')

# Create folders if needed
dir.create(save_dir, recursive = TRUE)
dir.create(qsave_dir, recursive = TRUE)
dir.create(qc_dir, recursive = TRUE)
dir.create(csv_dir, recursive = TRUE)
dir.create(plot_dir, recursive = TRUE)
```

## Sample list set
```{r Sample list setup, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Set sample list name = index, item = samples
samples_named <- 
  list(
    "FF061" = c("SOM333", "SOM318", "SOM303", "SOM301"),
    "FF062" = c("SOM326", "SOM325", "SOM323", "SOM322"),
    "FF063" = c("SOM329", "SOM328", "SOM327", "SOM302"),
    "FF064" = c("SOM378", "SOM377", "SOM376", "P07"))

# Set sample list without names
samples <- 
  names(seurat_list)

# Sample Info
info <- list(
  "SOM301" = c("P23", "cKO"),
  "SOM303" = c("P23", "cKO"),
  "SOM318" = c("P08", "cKO"),
  "SOM333" = c("P23", "Control"),
  "SOM322" = c("P56", "Control"),
  "SOM323" = c("P56", "Control"),
  "SOM325" = c("P56", "cKO"),
  "SOM326" = c("P56", "cKO"),
  "SOM302" = c("P23", "Control"),
  "SOM327" = c("P08", "Control"),
  "SOM328" = c("P08", "Control"),
  "SOM329" = c("P08", "cKO"),
  "SOM376" = c("P00", "cKO"),
  "SOM377" = c("P00", "Control"),
  "SOM378" = c("P00", "Control"),
  "P07" = c("P00", "cKO")
)
```

# Initial processing

## Load data files
```{r [O] Load RNA data from H5 files, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/01__load_rna_from_h5.R")
rna_list <- 
  load_rna_from_h5(h5_input_dir,
                   samples_named,
                   save = T,
                   save_dir = qsave_dir)

# Save
qs::qsave(rna_list, 
          file = "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/QSAVE/01__rna_list.qs")

```

## Seurat Object creation
```{r [O] Create Seurat Object from RNA data, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/02__create_seurat_object.R")

# load previous object if needed
# rna_list <- qs::qread("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/QSAVE/01__rna_list.qs")

seurat_list <- 
  create_seurat_object.R(rna_list,
                       h5_input_dir, # for meta data file path
                       samples_named,
                       project = 'SLC35A2Olig',
                       min.cell = 3,
                       min.feature = 200,
                       save = T,
                       save_dir = qsave_dir)

# Raw count matrices loaded
# Seurat objects created with  metadata
# Sample/index identifiers added
# Mitochondrial gene percentages calculated

# Save
qs::qsave(seurat_list, 
          file = file.path(qsave_dir, "002__rna_list_seurat.qs"))
```

## Initial Vizualization
```{r [V] Initial Visual Overview , echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# R script located at "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/02_Initial_VlnPlot.R"
plots <- qc_VlnPlot(seurat_list, save_dir = plot_dir)
```

## Basic metrics
```{r [S] Initial process - Stats Table, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# From the literature
# Cell-Level Filtering:
## UMI counts: 1000 < UMI < 25,000
## Gene features: > 400 genes (some use > 200)
## Mitochondrial %: < 15%
# Gene-Level Filtering:
## Min cells expressing: â‰¥ 3 cells
## Remove mitochondrial genes for downstream analysis

# Create table with counts information
SummaryTable <- data.frame()

for (sample in seurat_list) {
  dt <- data.frame(
    # Sample name
    MouseID = unique(sample$MouseID),
    
    TotalCells = nrow(sample@meta.data),
    
    nCount_mean = mean(sample$nCount_RNA),
    nFeature_RNA_mean = mean(sample$nFeature_RNA),
    percent_mt_rna_mean = mean(sample$percent_mt_rna),
    
    nCount_median = median(sample$nCount_RNA),
    nFeature_RNA_median = median(sample$nFeature_RNA),
    percent_mt_rna_median = median(sample$percent_mt_rna),
    
    nCount_min = min(sample$nCount_RNA),
    nCount_max = max(sample$nCount_RNA),
    
    nFeature_RNA_min = min(sample$nFeature_RNA),
    nFeature_RNA_max = max(sample$nFeature_RNA),
    
    percent_mt_rna_min = min(sample$percent_mt_rna),
    percent_mt_rna_max = max(sample$percent_mt_rna)
  )
  
  # Put all in one table
  SummaryTable <- bind_rows(SummaryTable, dt)
}

write.csv(SummaryTable, file = file.path(csv_dir, "Initial_stats.csv"))

#### Same table but using map_dfr  #### 
SummaryTable <- map_dfr(names(seurat_list), ~{
  meta <- seurat_list[[.x]]@meta.data
  data.frame(
    # Sample name
    MouseID = unique(meta$MouseID),

    TotalCells = nrow(meta),

    nCount_mean = mean(meta$nCount_RNA),
    nFeature_RNA_mean = mean(meta$nFeature_RNA),
    percent_mt_rna_mean = mean(meta$percent_mt_rna),

    nCount_median = median(meta$nCount_RNA),
    nFeature_RNA_median = median(meta$nFeature_RNA),
    percent_mt_rna_median = median(meta$percent_mt_rna),

    nCount_min = min(meta$nCount_RNA),
    nCount_max = max(meta$nCount_RNA),

    nFeature_RNA_min = min(meta$nFeature_RNA),
    nFeature_RNA_max = max(meta$nFeature_RNA),

    percent_mt_rna_min = min(meta$percent_mt_rna),
    percent_mt_rna_max = max(meta$percent_mt_rna))
})

```

## QC metric exprlore
```{r [S] Initial process - QC threshold test, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/03_qc_threshold_test.R')
QC_test <- qc_threshold(seurat_list = seurat_list,
                         min_count = 1000,
                         max_count = 25000,
                         min_feature = 400,
                         max_mt = 5)
```

# Filtering and Processing
## Filter based on QC matrics
```{r [O] Initial process - Filter, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/03__rna_qc.R")

# load previous object if needed
# seurat_list <- qs::qread("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/QSAVE/002__rna_list_seurat.qs")

# Filter the cells
filtered_list <- rna_qc(
  seurat_list,
  samples,
  nCount_RNA_low = 1000,
  nCount_RNA_high = 25000,
  nFeature_RNA_low = 400,
  mt_rna = 5,
  save = T,
  save_dir_qs = qsave_dir,
  save_dir_qc = qc_dir)

# Check filtered cell counts
QC_summary <- read_csv("WIP/QC/QC_summary.csv")

# Save
qs::qsave(filtered_list, 
          file = file.path(qsave_dir, "03__filtered_rna_list.qs"))

```

## Pre-processing
```{r [O] Normalization and preprocessing, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/04__process_rna.R")

# Read previous object if needed.
# filtered_list <- qs::qread(file.path(qsave_dir, "03__filtered_rna_list.qs"))

# Normalization, SCTrancform, PCA and UMAP/TSNE
processed_list <- process_rna(
  filtered_list,
  save = TRUE,
  save_dir = qsave_dir)

# Save
qs::qsave(processed_list, 
          file = file.path(qsave_dir, '04__processed_rna_list'))
```

# Integration
Integration with SCT method -> failed due to limit of data size can handeled by Seurat

## Integration with RPCA method
Integration with SCT can't cover the number of cells in our object so trying RPCA method in next section
```{r [O] Integrate RNA objects with RPCA, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/05__rpca_rna_integration.R')

# Read previous object if needed.
# processed_list <- qs::qread(file.path(qsave_dir, "04__processed_rna_list"))

# Integration
integrated <- rpca_integration(processed_list, 
                               save = TRUE,
                               save_dir = qsave_dir)

```

# Process Integrated object
## PCA, Dimentional Reduction
```{r [O] Process integrated object, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/06__process_integrated.R')

# Read previous object if needed.
# integrated <- qs::qread(file.path(qsave_dir, "05__integrated_rna.qs"))

# Process integrated object
integrated <- process_integrated(integrated,
                                 save = TRUE,
                                 save_dir = qsave_dir,
                                 )

```

### Initial visualization - DimPlot
```{r [V] Initial DimPlot after integration, message=TRUE, warning=TRUE, paged.print=TRUE}
# set assay and Idents
Seurat::DefaultAssay(integrated) <- "RNA"
Seurat::Idents(integrated) <- "MouseID"

# DimPLot
dp <- Seurat::DimPlot(integrated, reduction = 'umap', group.by = 'TimePoint', alpha = 0.3) |
  Seurat::DimPlot(integrated, reduction = 'umap', group.by = 'GenoType', alpha = 0.3) |
  Seurat::DimPlot(integrated, reduction = 'umap', group.by = "MouseID", alpha = 0.3) |
  Seurat::DimPlot(integrated, reduction = 'umap', group.by = "SampleID", alpha = 0.3)

# Save
ggplot2::ggsave(filename = file.path(plot_dir, "DimPlot_umap_initial_all.png"), dp, width = 18, height = 4, dpi = 300)

Seurat::DimPlot(integrated, reduction = 'umap', split.by = 'TimePoint', alpha = 0.3)
Seurat::DimPlot(integrated, reduction = 'umap', split.by = 'GenoType', alpha = 0.3)
Seurat::DimPlot(integrated, reduction = 'umap', split.by = "MouseID", alpha = 0.3)
Seurat::DimPlot(integrated, reduction = 'umap', split.by = "SampleID", alpha = 0.3)
```

## Cluster and Markers, additional meta data
```{r [O] Clusteriazation, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/07__clustering.R')

# Read previous object if needed.
# integrated <- qs::qread(file.path(qsave_dir, "06__integrated_rna.qs"))

integrated <- clustering(integrated,
                         range_start = 0.4,
                         range_end = 1.2,
                         range_step = 0.1,
                         save = TRUE,
                         qsave_dir = qsave_dir,
                         markers_dir = file.path(csv_dir, "Markers"))


# Additional meta data
# Forgot to add this early in the process so adding now.
# Convert info list to lookup vectors
timepoint_lookup <- sapply(info, function(x) x[1])
genotype_lookup <- sapply(info, function(x) x[2])

# Add to metadata
integrated@meta.data <- integrated@meta.data %>%
  mutate(
    TimePoint = timepoint_lookup[MouseID],
    GenoType = genotype_lookup[MouseID]
  )

# Add additional meta data combine age and genotype
integrated@meta.data$TimeGeno <- paste0(integrated@meta.data$TimePoint, "_", integrated@meta.data$GenoType)

# Join layers after integration in RNA assay
Seurat::DefaultAssay(integrated) <- "RNA"
integrated <- SeuratObject::JoinLayers(integrated)

# Save
qs::qsave(integrated, file = file.path(qsave_dir, "08__clustered_metadata_added.qs"))
```

# Cell type annotation
## Use Mousebrain.org object as reference for cell typing. Using Jesse's custom object.
```{r [O] Cell type, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/08_ref_mousebrain_annotation.R')

# Read previous object if needed.
# integrated <- qs::qread(file.path(qsave_dir, "08__clustered_metadata_added.qs"))

# Read reference object for cell typing
ref_obj <- readRDS("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/mouseatlas_cut_processed.rds")

Seurat::DefaultAssay(integrated) <- "SCT"
Seurat::DefaultAssay(ref_obj) <- "SCT"

# run mouse annotation
integrated <- ref_mousebrain_annotation(ref_obj,
                                        integrated,
                                        save = T,
                                        save_dir = qsave_dir)

```

## Clean
```{r [O] Clean ref annotation, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated <- qs::qread(file.path(qsave_dir, "09__ref_annotated_obj.qs"))

# Ref cell type prediction will be stored in oligsub_man2$predicted.id.
Seurat::Idents(integrated) <- integrated$predicted.id

# Visualize
Seurat::DimPlot(integrated, reduction = "umap", label = T, repel = T) + NoLegend()

# get names and shorten them
ln <- unique(integrated$predicted.id)

sn <- c("Ex_Cortex", "Microglia", "Ex_CA3", "NPC", "Ex_CA1", "Interneuron", "Astrocytes", "Granule_DG", "Microglia",
        "MFOL", "Vascular", "Interneuron", "COP", "Interneuron", "Interneuron", "CGE", "Ependymal", "CGE",
        "Mature_OL", "OPC", "Cajal_Retzius", "Interneuron", "Pericytes", "Interneuron", "RGL_DG", "Astrocytes", "Granule_Blast_DG",
        "Macrophages", "Vascular","MGE", "NFOL", "Trilaminar", "Vascular", "Vascular", "Interneuron", "Vascular", "Interneuron")

# Mapping sn-ln
name_map <- setNames(sn, ln)

# Function to replace long names to short names
shortened <- function(name) {ifelse(name %in% names(name_map), name_map[name], name)}

# Apply function
integrated$ref_short <- shortened(integrated$predicted.id)

# Check UMAP
Seurat::Idents(integrated) <- "ref_short"
dp_ref <- Seurat::DimPlot(integrated, reduction = "umap", label = T, repel = T) + Seurat::NoLegend()

# Also, cleaned out metadata. all prediction score were deleted.
# #  all prediction score were deleted.
integrated@meta.data <- integrated@meta.data[, !grepl(pattern = "prediction.score", x = colnames(integrated@meta.data))]

# Save
ggplot2::ggsave(dp_ref, file = file.path(plot_dir, "DimPlot_mousebrain_ref.png"), width = 8, height = 8, dpi = 300)
qs::qsave(integrated, file = file.path(qsave_dir, "09__ref_annotated_obj.qs"))
```

## Milo try
```{r [O] Milo for identifying cluster between cKO vs Control, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
## ~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/99_Milo.R
# https://www.nature.com/articles/s41587-021-01033-z
```

## Gene list
```{r [S] Gene list, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/99_gene_list.R"
```

## Manual Cell type check
```{r [V] Cell type check with marker genes, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/99_CellTypeWork.R"

Seurat::Idents(integrated) <- "CellTypes" 
Seurat::DimPlot(integrated, reduction = "umap", label = TRUE, repel = TRUE) + Seurat::NoLegend()
```

## Visualize Cell tyep markers
```{r [V]] DotPlot, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object
# integrated <- qs::qread(file.path(qsave_dir, "10__metadata_edited_obj.qs"))

# cell type vectors
ct <- c('Astrocytes', 'Microglia', 'Ependymal', 'ExN', 'InhN', 'NPC', 'OPC_1', 'OPC_2', 'COP/NFOL', 'MFOL', 'Mature_OL', 'Pericytes', 'Vascular')

# Order Cell types
integrated$CellTypes <- factor(integrated$CellTypes, levels = ct)

# Set marker genes to plot
markers <- c('Aqp4', 'Aldh1l1', 'Gfap', 'P2ry12', 'Aif1', 'Tmem212', 'Foxj1', 'Slc17a7', 'Gad1','Cux1', 'Nes', 'Pdgfra', 'Cspg4', 'Enpp6', 'Plp1', 'Mog', 'Vtn', 'Pdgfrb')

# Create DotPlot
dp <- Seurat::DotPlot(integrated, 
        features = markers, 
        group.by = "CellTypes",
        cols = c("darkgrey", "indianred1")) +
  ggplot2::coord_flip() +
  ggplot2::theme(
        axis.text.x =
            ggplot2::element_text(
                angle = 90,
                vjust = 0.8,
                hjust = 1,
                size = 10
            ),
        axis.text.y =
            ggplot2::element_text(
                size = 10
            ),
        axis.title.x =
            ggplot2::element_blank(),
        axis.title.y =
            ggplot2::element_blank(),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8)
    )

# Additional DimPlot
dimp <- SeuratExtend::DimPlot2(integrated, 
                               reduction = 'umap', 
                               group.by = 'CellTypes',
                               label = TRUE,
                               label.size = 5,
                               repel = TRUE,
                               theme = list(
                                 Seurat::NoLegend(), 
                                 Seurat::NoAxes(), 
                                 ggtitle("UMAP by Cell Type")
                                 )) +  
  theme_umap_arrows(x_label = "UMAP_1",
                    y_label = "UMAP_2") 


# Save
ggplot2::ggsave(dp, filename = file.path(plot_dir, "DotPlot_CellMarkers.png"), width = 7, height = 7, dpi = 300)
ggplot2::ggsave(dimp, filename = file.path(plot_dir, "DimPlot_CellType.png"), width = 7, height = 7, dpi = 300)
```

# SLC35A2 expression check
```{r [V] Check SLC35A2 expression, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Assay and idents set
Seurat::DefaultAssay(integrated) <- "SCT"
Seurat::Idents(integrated) <- "ref_short"

# Looking only Olig cells
oligsub <- subset(integrated, integrated$ref_short %in% c("OPC", "COP", "NFOL", "MFOL", "Mature_OL"))
oligsub$ref_short <- factor(oligsub$ref_short, levels = c("OPC", "COP", "NFOL", "MFOL", "Mature_OL"))

# Assay and idents set
Seurat::DefaultAssay(oligsub) <- "SCT"
Seurat::Idents(oligsub) <- "MouseID"

# Visualize
SeuratExtend::VlnPlot2(oligsub, features = "Slc35a2",show.mean = F, pt = F, box = F)
Seurat::FeaturePlot(integrated, features = "Slc35a2", label = F, repel = T, split.by = "TimeGeno")

# Save for refrence
ggplot2::ggsave(slc35a2, file = file.path(plot_dir, "Feature_slc35a2.png"), width = 32, height = 8, dpi = 300)

# Slc35a2 expression doesn't seem to be reduced in cKO cells. it might be due to ambient RNA from dissociateion/library prep. Plan to implement SoupX (Tracy will provide the script to start). Then I can compare once again. 
# SoupX didn't change much.
## Nuclear RNA only, 
```

# Cell count check
```{r [S] Cell count Summary, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
## "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/99_CellCount_Slc35a2Check.R"
# Done cell count summary, Slc35a2 expressing cell check
# plot save in "./WIP/Plots/Cell_counts
```

# Doublet process
## Prepare list of objects.
```{r [O] Add cluster info to individual object, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object
# integrated <- qs::qread(file.path(qsave_dir, "10__metadata_edited_obj.qs"))
# processed_list <- qs::qread(file.path(qsave_dir, "04__processed_rna_list"))

# Transfer cluster identity
meta <- integrated@meta.data %>%
  tibble::rownames_to_column("CellID") %>%
  select(CellID, ref_short, CellTypes)

for (i in seq_along(processed_list)) {
  obj_meta <- processed_list[[i]]@meta.data %>%
    tibble::rownames_to_column("CellID")
  
  new_meta <- obj_meta %>%
    left_join(meta, by = "CellID")%>%
    tibble::column_to_rownames("CellID")
  
  processed_list[[i]]@meta.data <- new_meta
    
}

# save
qs::qsave(processed_list, file = file.path(qsave_dir, "04_processed_rna_list_id.qs"))
```

## DoubletFinder
```{r [O] Doublet filter check, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Run doubletfinder
# Read previous object
# processed_list <- qs::qread(file.path(qsave_dir, "04_processed_rna_list_id.qs"))
# integrated <- qs::qread(file.path(qsave_dir, "10__metadata_edited_obj.qs"))

# Run
processed_list <- find_dbl(processed_list, save = TRUE, save_dir = qsave_dir)

# Collect all metadata from processed_list
all_individual_meta <- list()

for (i in seq_along(processed_list)) {
  individual_meta <- processed_list[[i]]@meta.data %>%
    tibble::rownames_to_column("CellID") %>%
    select(CellID, contains("Doublet_Status")) # Select doublet-related columns
    
  all_individual_meta[[i]] <- individual_meta
}

# Combine all individual metadata
combined_individual_meta <- do.call(bind_rows, all_individual_meta)

# Transfer back to integrated object
integrated_meta <- integrated@meta.data %>%
  tibble::rownames_to_column("CellID") %>%
  left_join(combined_individual_meta, by = "CellID") %>%
  tibble::column_to_rownames("CellID")

# Update integrated object metadata
integrated@meta.data <- integrated_meta

# save
qs::qsave(integrated, file = file.path(qsave_dir, "11__dbl_added.qs"))
```

### Doublet counts
```{r [S] Doublet Stats, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object
# integrated <- qs::qread(file.path(qsave_dir, "11__dbl_added.qs"))
doublet_ratio <- round(table(integrated$Doublet_Status)[[1]]/table(integrated$Doublet_Status)[[2]],3)

doublet_grouped <- integrated@meta.data %>%
  group_by(ref_short) %>%
  count(Doublet_Status)
```

# Olig subset
Proper subset and integration process
## Subset
```{r [O] Subset Olig, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object
processed_list <- qs::qread(file.path(qsave_dir, "04_processed_rna_list_dbl.qs"))
# integrated <- qs::qread(file.path(qsave_dir, "11__dbl_added.qs"))

# Cell clusters
oligos <- c("OPC_1", "OPC_2", "COP/NFOL", "MFOL", "Mature_OL")
oligos_npc <- c(oligos, "NPC")

# Subset based on manual
processed_list_man <- processed_list
for (i in seq_along(processed_list_man)) {
  processed_list_man[[i]] <- subset(processed_list_man[[i]], processed_list_man[[i]]$CellTypes %in% oligos)
}

# Subset based on manual
processed_list_ref <- processed_list
for (i in seq_along(processed_list_man)) {
 processed_list_ref[[i]] <- subset(processed_list_man[[i]], processed_list_man[[i]]$ref_short %in% c("OPC", "COP", "NFOL", "MFOL", "Mature_OL"))
}

# Subset based on manual
processed_list_man_npc <- processed_list
for (i in seq_along(processed_list_man_npc)) {
  processed_list_man_npc[[i]] <- subset(processed_list_man_npc[[i]], processed_list_man_npc[[i]]$CellTypes %in% oligos_npc)
}

# Subset based on manual
processed_list_ref_npc <- processed_list
for (i in seq_along(processed_list_ref_npc)) {
 processed_list_ref_npc[[i]] <- subset(processed_list_ref_npc[[i]], processed_list_ref_npc[[i]]$ref_short %in% c("OPC", "COP", "NFOL", "MFOL", "Mature_OL", "NPC"))
}

# Save
qs::qsave(processed_list_ref, file = file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_ref.qs"))
qs::qsave(processed_list_man, file = file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_man.qs"))
qs::qsave(processed_list_ref_npc, file = file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_ref_npc.qs"))
qs::qsave(processed_list_man_npc, file = file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_man_npc.qs"))
```

## Integration
### Integration with RPCA method
Integration with SCT can't cover the number of cells in our object so trying RPCA method in next section
```{r [O] Integrate RNA objects with RPCA, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source code changed for RunPCA due to low cell number in ref obj (P07b has 49 cells). default npcs are 50 and errored out so changed npcs to 48.
# source("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/05__rpca_rna_integration.R")

# Read previous object if needed.
# processed_list_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_ref.qs"))
# processed_list_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_man.qs"))
# processed_list_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_ref_npc.qs"))
# processed_list_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "04__processed_rna_list_subset_man_npc.qs"))

# Integration
integrated_ref <- rpca_integration(processed_list_ref, 
                               save = T,
                               suffix = "_ref",
                               npc = 48,
                               save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_man <- rpca_integration(processed_list_man, 
                               save = T,
                               suffix = "_man",
                               npc = 48,
                               save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_ref_npc <- rpca_integration(processed_list_ref_npc, 
                               save = T,
                               suffix = "_ref_npc",
                               npc = 48,
                               save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_man_npc <- rpca_integration(processed_list_man_npc, 
                               save = T,
                               suffix = "_man_npc",
                               npc = 48,
                               save_dir = file.path(qsave_dir, "OligoSubset"))

```

## Process Integrated object
### PCA, Dimentional Reduction
```{r [O] Process integrated object, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/06__process_integrated.R')

# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "05__integrated_ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "05__integrated__man.qs"))
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "05__integrated_ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "05__integrated_man_npc.qs"))

# Process integrated object
integrated_ref <- process_integrated(integrated_ref,
                                 save = T,
                                 suffix = "_ref",
                                 save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_man <- process_integrated(integrated_man, 
                               save = T,
                               suffix = "_man",
                               save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_ref_npc <- process_integrated(integrated_ref_npc,
                                 save = T,
                                 suffix = "_man_npc",
                                 save_dir = file.path(qsave_dir, "OligoSubset"))

integrated_man_npc <- process_integrated(integrated_man_npc, 
                               save = T,
                               suffix = "_ref_npc",
                               save_dir = file.path(qsave_dir, "OligoSubset"))

```

## Cluster and Markers, additional meta data
```{r [O] Clusteriazation, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Change save name in source code!!
# source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/07__clustering.R')

# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "06__integrated_ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "06__integrated_man.qs"))
# integrated_ref_npc<- qs::qread(file.path(qsave_dir, "OligoSubset", "06__integrated_ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "06__integrated_man_npc.qs"))

# Join layers after integration in RNA assay
Seurat::DefaultAssay(integrated_man) <- "RNA"
integrated_man <- SeuratObject::JoinLayers(integrated_man)

integrated_man <- clustering(integrated_man,
                         range_start = 0.4,
                         range_end = 1.2,
                         range_step = 0.1,
                         save = TRUE,
                         suffix = "_man",
                         qsave_dir = file.path(qsave_dir, "OligoSubset"),
                         markers_dir = file.path(csv_dir, "Markers_OligoSubset"))


# Join layers after integration in RNA assay
Seurat::DefaultAssay(integrated_ref) <- "RNA"
integrated_ref <- SeuratObject::JoinLayers(integrated_ref)

integrated_ref <- clustering(integrated_ref,
                         range_start = 0.4,
                         range_end = 1.2,
                         range_step = 0.1,
                         save = TRUE,
                         suffix = "_ref",
                         qsave_dir = file.path(qsave_dir, "OligoSubset"),
                         markers_dir = file.path(csv_dir, "Markers_OligoSubset"))

# Join layers after integration in RNA assay
Seurat::DefaultAssay(integrated_man_npc) <- "RNA"
integrated_man_npc <- SeuratObject::JoinLayers(integrated_man_npc)

integrated_man_npc <- clustering(integrated_man_npc,
                         range_start = 0.4,
                         range_end = 1.2,
                         range_step = 0.1,
                         save = TRUE,
                         suffix = "_man_npc",
                         qsave_dir = file.path(qsave_dir, "OligoSubset"),
                         markers_dir = file.path(csv_dir, "Markers_OligoSubset"))

# Join layers after integration in RNA assay
Seurat::DefaultAssay(integrated_ref_npc) <- "RNA"
integrated_ref_npc <- SeuratObject::JoinLayers(integrated_ref_npc)

integrated_ref_npc <- clustering(integrated_ref_npc,
                         range_start = 0.4,
                         range_end = 1.2,
                         range_step = 0.1,
                         save = TRUE,
                         suffix = "_ref_npc",
                         qsave_dir = file.path(qsave_dir, "OligoSubset"),
                         markers_dir = file.path(csv_dir, "Markers_OligoSubset"))
```

```{r [O] Add metadata, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Additional meta data
# Forgot to add this early in the process so adding now.
# Convert info list to lookup vectors
timepoint_lookup <- sapply(info, function(x) x[1])
genotype_lookup <- sapply(info, function(x) x[2])

# Add to metadata
integrated_ref@meta.data <- integrated_ref@meta.data %>%
  mutate(
    TimePoint = timepoint_lookup[MouseID],
    GenoType = genotype_lookup[MouseID]
  )

# Add additional meta data combine age and genotype
integrated_ref@meta.data$TimeGeno <- paste0(integrated_ref@meta.data$TimePoint, "_", iintegrated_ref@meta.data$GenoType)


# Add to metadata
integrated_man@meta.data <- integrated_man@meta.data %>%
  mutate(
    TimePoint = timepoint_lookup[MouseID],
    GenoType = genotype_lookup[MouseID]
  )

# Add additional meta data combine age and genotype
integrated_man@meta.data$TimeGeno <- paste0(integrated_man@meta.data$TimePoint, "_", integrated_man@meta.data$GenoType)

# Add to metadata
integrated_ref_npc@meta.data <- integrated_ref_npc@meta.data %>%
  mutate(
    TimePoint = timepoint_lookup[MouseID],
    GenoType = genotype_lookup[MouseID]
  )

# Add additional meta data combine age and genotype
integrated_ref_npc@meta.data$TimeGeno <- paste0(integrated_ref_npc@meta.data$TimePoint, "_", integrated_ref_npc@meta.data$GenoType)

# Add to metadata
integrated_man_npc@meta.data <- integrated_man_npc@meta.data %>%
  mutate(
    TimePoint = timepoint_lookup[MouseID],
    GenoType = genotype_lookup[MouseID]
  )

# Add additional meta data combine age and genotype
integrated_man_npc@meta.data$TimeGeno <- paste0(integrated_man_npc@meta.data$TimePoint, "_", integrated_man_npc@meta.data$GenoType)

# Save
qs::qsave(integrated_ref, file = file.path(qsave_dir, "OligoSubset", "08__clustered_ref.qs"))
qs::qsave(integrated_man, file = file.path(qsave_dir, "OligoSubset", "08__clustered_man.qs"))
qs::qsave(integrated_ref_npc, file = file.path(qsave_dir, "OligoSubset", "08__clustered_ref_npc.qs"))
qs::qsave(integrated_man_npc, file = file.path(qsave_dir, "OligoSubset", "08__clustered_man_npc.qs"))
```

### Initial visualization - DimPlot

```{r [V] Initial DimPlot glimpse, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_man.qs"))
# integrated_ref_npc<- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_man_npc.qs"))

# set assay and Idents
Seurat::DefaultAssay(iintegrated_ref) <- "RNA"
Seurat::Idents(iintegrated_ref) <- "MouseID"

Seurat::DefaultAssay(integrated_man) <- "RNA"
Seurat::Idents(integrated_man) <- "MouseID"

Seurat::DefaultAssay(iintegrated_ref_npc) <- "RNA"
Seurat::Idents(iintegrated_ref_npc) <- "ref_short"

Seurat::DefaultAssay(integrated_man_npc) <- "RNA"
Seurat::Idents(integrated_man_npc) <- "CellTypes"

# DimPLot
dp_ref <- Seurat::DimPlot(integrated_ref, reduction = 'umap', group.by = 'TimePoint', alpha = 0.3) |
  Seurat::DimPlot(integrated_ref, reduction = 'umap', group.by = 'GenoType', alpha = 0.3) |
  Seurat::DimPlot(integrated_ref, reduction = 'umap', group.by = "MouseID", alpha = 0.3) |
  Seurat::DimPlot(integrated_ref, reduction = 'umap', group.by = "SampleID", alpha = 0.3)

dp_man <- Seurat::DimPlot(integrated_man, reduction = 'umap', group.by = 'TimePoint', alpha = 0.3) |
  Seurat::DimPlot(integrated_man, reduction = 'umap', group.by = 'GenoType', alpha = 0.3) |
  Seurat::DimPlot(integrated_man, reduction = 'umap', group.by = "MouseID", alpha = 0.3) |
  Seurat::DimPlot(integrated_man, reduction = 'umap', group.by = "SampleID", alpha = 0.3)

dp_ref_npc <- Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', group.by = 'TimePoint', alpha = 0.3) |
  Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', group.by = 'GenoType', alpha = 0.3) |
  Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', group.by = "MouseID", alpha = 0.3) |
  Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', group.by = "SampleID", alpha = 0.3)

dp_man_npc <- Seurat::DimPlot(integrated_man_npc, reduction = 'umap', group.by = 'TimePoint', alpha = 0.3) |
  Seurat::DimPlot(integrated_man_npc, reduction = 'umap', group.by = 'GenoType', alpha = 0.3) |
  Seurat::DimPlot(integrated_man_npc, reduction = 'umap', group.by = "MouseID", alpha = 0.3) |
  Seurat::DimPlot(integrated_man_npc, reduction = 'umap', group.by = "SampleID", alpha = 0.3)

Seurat::DimPlot(integrated_ref, reduction = 'umap', split.by = 'TimePoint', alpha = 0.3)
Seurat::DimPlot(integrated_ref, reduction = 'umap', split.by = 'GenoType', alpha = 0.3)
Seurat::DimPlot(integrated_ref, reduction = 'umap', split.by = "MouseID", alpha = 0.3)
Seurat::DimPlot(integrated_ref, reduction = 'umap', split.by = "SampleID", alpha = 0.3)

Seurat::DimPlot(integrated_man, reduction = 'umap', split.by = 'TimePoint', alpha = 0.3)
Seurat::DimPlot(integrated_man, reduction = 'umap', split.by = 'GenoType', alpha = 0.3)
Seurat::DimPlot(integrated_man, reduction = 'umap', split.by = "MouseID", alpha = 0.3)
Seurat::DimPlot(integrated_man, reduction = 'umap', split.by = "SampleID", alpha = 0.3)

Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = 'TimePoint', alpha = 0.3)
Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3)
Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = "MouseID", alpha = 0.3)
Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = "SampleID", alpha = 0.3)

Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = 'TimePoint', alpha = 0.3)
Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3)
Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = "MouseID", alpha = 0.3)
Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = "SampleID", alpha = 0.3)

ref <- Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3, group.by = "ref_short", label = T, repel = T) + NoLegend() + ggtitle("ref") |
Seurat::DimPlot(integrated_ref_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3, group.by = "CellTypes", label = T, repel = T) + NoLegend() + ggtitle("manual")

man <- Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3, group.by = "ref_short", label = T, repel = T) + NoLegend() + ggtitle("ref") |
Seurat::DimPlot(integrated_man_npc, reduction = 'umap', split.by = 'GenoType', alpha = 0.3, group.by = "CellTypes", label = T, repel = T) + NoLegend() + ggtitle("manual")

ref / man
```

## Cell type annotation
Use Mousebrain.org object as reference for cell typing. Using Jesse's custom object.
```{r [O] Cell type, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
source('~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/08__ref_mousebrain_annotation.R')

# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_man.qs"))
# integrated_ref_npc<- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "08__clustered_man_npc.qs"))

# Read reference object for cell typing
ref_obj <- readRDS("~/SLC35A2_Olig2cKO_snRNA_JY/WIP/mouseatlas_cut_processed.rds")

Seurat::DefaultAssay(integrated_ref) <- "SCT"
Seurat::DefaultAssay(integrated_man) <- "SCT"
Seurat::DefaultAssay(integrated_ref_npc) <- "SCT"
Seurat::DefaultAssay(integrated_man_npc) <- "SCT"

# run mouse annotation
integrated_ref <- ref_mousebrain_annotation(ref_obj,
                                        integrated_ref,
                                        save = T,
                                        suffix = "_ref",
                                        save_dir = file.path(qsave_dir, "OligoSubset")
                                        )

# run mouse annotation
integrated_man <- ref_mousebrain_annotation(ref_obj,
                                        integrated_man,
                                        save = T,
                                        suffix = "_man",
                                        save_dir = file.path(qsave_dir, "OligoSubset")
                                        )

# run mouse annotation
integrated_ref_npc <- ref_mousebrain_annotation(ref_obj,
                                        integrated_ref_npc,
                                        save = T,
                                        suffix = "_ref_npc",
                                        save_dir = file.path(qsave_dir, "OligoSubset")
                                        )

# run mouse annotation
integrated_man_npc <- ref_mousebrain_annotation(ref_obj,
                                        integrated_man_npc,
                                        save = T,
                                        suffix = "_man_npc",
                                        save_dir = file.path(qsave_dir, "OligoSubset")
                                        )
```

### Metadata cleaning
```{r [O] Clean out metadata., echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "09__ref_annotated_obj_ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "09__ref_annotated_obj_man.qs"))
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "09__ref_annotated_obj_ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "09__ref_annotated_obj_man_npc.qs"))

#  all prediction score were deleted.
integrated_ref@meta.data <- integrated_ref@meta.data[, !grepl(pattern = "prediction.score", x = colnames(integrated_ref@meta.data))]

integrated_ref_npc@meta.data <- integrated_ref_npc@meta.data[, !grepl(pattern = "prediction.score", x = colnames(integrated_ref_npc@meta.data))]

integrated_man@meta.data <- integrated_man@meta.data[, !grepl(pattern = "prediction.score", x = colnames(integrated_man@meta.data))]

integrated_man_npc@meta.data <- integrated_man_npc@meta.data[, !grepl(pattern = "prediction.score", x = colnames(integrated_man_npc@meta.data))]

# Save
qs::qsave(integrated_ref, file = file.path(qsave_dir, "OligoSubset", "10__ref.qs"))
qs::qsave(integrated_man, file = file.path(qsave_dir, "OligoSubset", "10__man.qs"))
qs::qsave(integrated_ref_npc, file = file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
qs::qsave(integrated_man_npc, file = file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))
```

### Clean cluster name
```{r [O] Clean ref annotation, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref.qs"))
# integrated_man <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man.qs"))
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

# Function to replace long names to short names
shortened <- function(name) {ifelse(name %in% names(name_map), name_map[name], name)}

# Ref cell type prediction will be stored in oligsub_man2$predicted.id.
Seurat::Idents(integrated_ref) <- integrated_ref$predicted.id
Seurat::Idents(integrated_man) <- integrated_man$predicted.id
Seurat::Idents(integrated_ref_npc) <- integrated_ref_npc$predicted.id
Seurat::Idents(integrated_man_npc) <- integrated_man_npc$predicted.id

# get names and shorten them - ref
ln <- unique(integrated_ref$predicted.id)
sn <- c("MFOL", "Mature_OL", "OPC", "COP", "NFOL", "Interneuron", "ExCA1", "ExCortex", "ExCA3", "NPC", "Granule_DG", "Vascular", "Vascular", "Granule_Blast_DG", "Cajal_Retzius")
name_map <- setNames(sn, ln)
integrated_ref$ref_short2 <- shortened(integrated_ref$predicted.id)

# get names and shorten them - man
ln <- unique(integrated_man$predicted.id)
sn <- c("ExCA3", "ExCortex", "MFOL", "Mature_OL", "OPC", "COP", "Interneuron", "NFOL", "Interneuron", "MGE", "Granule_DG", "ExCA1", "Interneuron", "Cajal_Retzius", "CGE", "NPC", "Interneuron", "Vascular", "Vascular", "Granule_Blast_DG")
name_map <- setNames(sn, ln)
integrated_man$ref_short2 <- shortened(integrated_man$predicted.id)

# get names and shorten them - ref_npc
ln <- unique(integrated_ref_npc$predicted.id)
sn <- c("Vascular", "MFOL", "COP", "ExCortex", "Mature_OL", "OPC", "Microglia", "NFOL", "Interneuron", "Microglia", "Astrocytes", "NPC", "ExCA1", "ExCA3", "Granule_Blast_DG", "Cajal_Retzius", "Granule_DG", "Macrophages", "Interneuron", "Astrocytes", "Ependymal", "Vascular", "Interneuron", "RGL", "Vascular")
name_map <- setNames(sn, ln)
integrated_ref_npc$ref_short2 <- shortened(integrated_ref_npc$predicted.id)

# get names and shorten them - man_npc
ln <- unique(integrated_man_npc$predicted.id)
sn <- c("ExCA3", "ExCortex", "MFOL", "Mature_OL", "OPC", "COP", "Interneuron", "NFOL", "Interneuron", "NPC", "MGE", "Granule_DG", "ExCA1", "Cajal_Retzius", "Interneuron", "Astrocytes", "CGE", "RGL", "Granule_Blast_DG", "Interneuron", "Vascular", "Vascular", "Astrocytes")
name_map <- setNames(sn, ln)
integrated_man_npc$ref_short2 <- shortened(integrated_man_npc$predicted.id)

# Check UMAP
Seurat::Idents(integrated_ref) <- "ref_short2"
Seurat::Idents(integrated_man) <- "ref_short2"
Seurat::Idents(integrated_ref_npc) <- "ref_short2"
Seurat::Idents(integrated_man_npc) <- "ref_short2"


Seurat::DimPlot(integrated_ref, reduction = "umap", label = T, repel = T) + Seurat::NoLegend() |
Seurat::DimPlot(integrated_man, reduction = "umap", label = T, repel = T) + Seurat::NoLegend() |
Seurat::DimPlot(integrated_ref_npc, reduction = "umap", label = T, repel = T) + Seurat::NoLegend() |
Seurat::DimPlot(integrated_man_npc, reduction = "umap", label = T, repel = T) + Seurat::NoLegend()


# Order Column
integrated_ref$ref_short2 <- factor(integrated_ref$ref_short2, levels = c("Interneuron", "Cajal_Retzius", "ExCortex", "ExCA1", "ExCA3", "Granule_DG", "Granule_Blast_DG", "OPC", "COP", "NFOL", "MFOL", "Mature_OL", "NPC", "Vascular"))

integrated_man$ref_short2 <- factor(integrated_man$ref_short2, levels = c("CGE", "MGE", "Interneuron", "Cajal_Retzius", "ExCortex", "ExCA1", "ExCA3", "Granule_DG", "Granule_Blast_DG", "OPC", "COP", "NFOL", "MFOL", "Mature_OL", "NPC", "Vascular"))

integrated_ref_npc$ref_short2 <- factor(integrated_ref_npc$ref_short2, levels = c("Astrocytes", "Microglia", "Macrophages", "Interneuron", "Cajal_Retzius", "Ependymal", "RGL", "ExCortex", "ExCA1", "ExCA3", "Granule_DG", "Granule_Blast_DG", "OPC", "COP", "NFOL", "MFOL", "Mature_OL", "NPC", "Vascular"))

integrated_man_npc$ref_short2 <- factor(integrated_man_npc$ref_short2, levels = c("Astrocytes", "CGE", "MGE", "Interneuron", "Cajal_Retzius", "RGL", "ExCortex", "ExCA1", "ExCA3", "Granule_DG", "Granule_Blast_DG", "OPC", "COP", "NFOL", "MFOL", "Mature_OL", "NPC", "Vascular"))

# Save
qs::qsave(integrated_ref, file = file.path(qsave_dir, "OligoSubset", "10__ref.qs"))
qs::qsave(integrated_man, file = file.path(qsave_dir, "OligoSubset", "10__man.qs"))
qs::qsave(integrated_ref_npc, file = file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
qs::qsave(integrated_man_npc, file = file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))
```

### Cell Counts 
```{r [O] Filter low number cell cluster, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

table(integrated_ref_npc$ref_short2)
table(integrated_ref_npc$CellTypes)
table(integrated_man_npc$ref_short2)
table(integrated_man_npc$CellTypes)
```

## Manual Cell type check
```{r [V] Cell type check with marker genes, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# "~/SLC35A2_Olig2cKO_snRNA_JY/WIP/SCRIPTS/SourceCode/99_CellTypeWork_oligsub.R"
Seurat::Idents(integrated_man_npc) <- "CellTypes2" 
Seurat::DimPlot(integrated_man_npc, reduction = "umap", label = TRUE, repel = TRUE) + Seurat::NoLegend()

Seurat::Idents(integrated_ref_npc) <- "CellTypes2" 
Seurat::DimPlot(integrated_ref_npc, reduction = "umap", label = TRUE, repel = TRUE) + Seurat::NoLegend()
```

### Visualize - work with NPC included obj
#### Markers list
```{r Marker assign, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

# Set marker genes to plot
markers <- c("Olig2", "Pdgfra", "Neu4", "Enpp6", "Mog", "Mag", "Mbp", 'P2ry12', 'Nes', 'Vim')
OligMarkers_selected <-  c("Olig2", "Pdgfra", "Neu4", "Enpp6", "Mog", "Mag", "Mbp")
```

#### DotPlot
```{r [V], echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

# Order Column
integrated_man_npc$CellTypes <- factor(integrated_man_npc$CellTypes, levels = c("OPC_1", "OPC_2", "COP/NFOL", "MFOL", "Mature_OL", "NPC"))

integrated_man_npc$CellTypes2 <- factor(integrated_man_npc$CellTypes2, levels = c("OPC_1", "OPC_2", "OPC_3", "COP/NFOL", "MFOL/Mature_OL", "NPC"))

integrated_ref_npc$CellTypes <- factor(integrated_ref_npc$CellTypes, levels = c("OPC_1", "OPC_2", "COP/NFOL", "MFOL", "Mature_OL", "NPC", "Astrocytes", "Microglia", "Ependymal", "ExN", "InhN", "Pericytes", "Vascular"))

integrated_ref_npc$CellTypes2 <- factor(integrated_ref_npc$CellTypes2, levels = c("OPC_1", "OPC_2", "COP", "NFOL", "MFOL", "Mature_OL", "NPC", "Microglia", "Vascular"))

# Save
qs::qsave(integrated_man_npc, file = file.path(qsave_dir, "OligoSubset", "10_man_npc.qs"))
qs::qsave(integrated_ref_npc, file = file.path(qsave_dir, "OligoSubset", "10_ref_npc.qs"))

# Create DotPlot
Seurat::DotPlot(integrated_ref_npc, features = markers, group.by = "CellTypes2",cols = c("darkgrey", "indianred1")) +
  ggplot2::coord_flip() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.8, hjust = 1, size = 10), 
                 axis.text.y = ggplot2::element_text(size = 10), 
                 axis.title.x = ggplot2::element_blank(), 
                 axis.title.y = ggplot2::element_blank(), 
                 legend.title = element_text(size = 10), 
                 legend.text = element_text(size = 8)) + 
  ggtitle("ref based cluster subset") |
  Seurat::DotPlot(integrated_man_npc, features = markers, group.by = "CellTypes2",cols = c("darkgrey", "indianred1")) +
  ggplot2::coord_flip() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.8, hjust = 1, size = 10), 
                 axis.text.y = ggplot2::element_text(size = 10), 
                 axis.title.x = ggplot2::element_blank(), 
                 axis.title.y = ggplot2::element_blank(), 
                 legend.title = element_text(size = 10), 
                 legend.text = element_text(size = 8)) + 
  ggtitle("manual cluster subset")

```

##### FeaturePlot
```{r [V], echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

# FeaturePlot
Seurat::FeaturePlot(integrated_ref_npc, features = OligMarkers_selected, label = T, repel = T)
Seurat::FeaturePlot(integrated_man_npc, features = OligMarkers_selected, label = T, repel = T)
```

#### DimPlots
```{r [V], echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))
 

# DimPlot
DimPlot(integrated_man_npc, reduction = "umap", label = T, repel = T, group.by = "CellTypes2") + NoLegend() + NoLegend() |
  DimPlot(integrated_ref_npc, reduction = "umap", label = T, repel = T, group.by = "CellTypes2") + NoLegend()


SeuratExtend::DimPlot2(integrated_ref_npc, reduction = 'umap', group.by = 'CellTypes', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("ref based cluster subset"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2") |
SeuratExtend::DimPlot2(integrated_man_npc, reduction = 'umap', group.by = 'CellTypes', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("manual cluster subset"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2")

SeuratExtend::DimPlot2(integrated_ref_npc, reduction = 'umap', group.by = 'CellTypes2', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("ref based cluster subset"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2") |
SeuratExtend::DimPlot2(integrated_man_npc, reduction = 'umap', group.by = 'CellTypes2', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("manual cluster subset"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2")

SeuratExtend::DimPlot2(integrated_man_npc, reduction = 'umap', group.by = 'CellTypes', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("manual based cluster subset1"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2") |
SeuratExtend::DimPlot2(integrated_man_npc, reduction = 'umap', group.by = 'CellTypes2', label = TRUE, label.size = 5, repel = TRUE, 
                       theme = list(Seurat::NoLegend(), Seurat::NoAxes(), ggtitle("manual cluster subset2"))) +  
  theme_umap_arrows(x_label = "UMAP_1",y_label = "UMAP_2")

```

#### all
```{r [V], echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

Idents(integrated_man_npc) <- "CellTypes2"
DimPlot(integrated_man_npc, label = T, repel = T) + NoLegend()
FeaturePlot(integrated_man_npc, features = OligMarkers_selected)
FeaturePlot(integrated_man_npc, features = markers)

Idents(integrated_ref_npc) <- "CellTypes2"
DimPlot(integrated_ref_npc, label = T, repel = T) + NoLegend()
FeaturePlot(integrated_ref_npc, features = OligMarkers_selected)
FeaturePlot(integrated_ref_npc, features = markers)
```

### PCA visualization check
```{r [V] PCA check , echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
# integrated_ref_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__ref_npc.qs"))
# integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))

red <- integrated_man_npc[["pca"]]
feats_used <- rownames(Loadings(integrated_man_npc, reduction = "pca"))  # pca features

mat <- GetAssayData(integrated_man_npc, assay = "integrated", slot = "scale.data")[feats_used, , drop = FALSE]

md <- integrated_man_npc@meta.data %>%
  slice(match(colnames(mat), rownames(integrated_man_npc@meta.data))) %>%
  as.data.frame()

p <- pca(mat, metadata = md, scale  = FALSE, removeVar = 0)


biplot(p, showLoadings = TRUE, labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
screeplot(p, axisLabSize = 18, titleLabSize = 22) <- 
eigencorplot(p, metavars = c('CellTypes2', 'ref_short2'))
pairsplot(p)
```

**"file.path(qsave_dir, "OligoSubset", "10__man_npc.qs")" is my final olig subset object

# Olig-subsetted object
## Cell type verification with FindMarkers
I have odd OPC_2, OPC_3 cluster so it needed to be sorted.
```{r Inspect markers, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
# Read previous object if needed.
integrated_man_npc <- qs::qread(file.path(qsave_dir, "OligoSubset", "10__man_npc.qs"))
findmarkerlist <- as.data.frame(read_csv("WIP/CSV/Markers_OligoSubset/Olig_cluster_man_npc_1.2_markers_man_npc.csv", col_select = -1))

# Clean meta data little
# integrated_man_npc$predicted.id <- NULL
# integrated_man_npc$TimePoint <- integrated_man_npc$TimePoint
# integrated_man_npc$GenoType <- integrated_man_npc$GenoType

FilteredList <- findmarkerlist %>%
  filter(p_val <= 0.05) %>%
  arrange(cluster, desc(p_val_adj), avg_log2FC) %>%
  group_by(cluster) %>%
  top_n(10) %>%
  arrange(cluster, desc(p_val_adj), avg_log2FC)

celltype2markers <- FindAllMarkers(integrated_man_npc, 
               assay = 'integrated',
               slot = 'data',
               only.pos = FALSE,
               min.pct = 0.25,
               logfc.threshold = 0.25,
               test.use = 'wilcox',
               group.by = "CellTypes2")  

filtered <- celltype2markers %>%
  filter(p_val <= 0.05) %>%
  arrange(cluster, desc(p_val_adj), avg_log2FC) %>%
  group_by(cluster) %>%
  top_n(10) %>%
  arrange(cluster, desc(p_val_adj), avg_log2FC)
```

### Based on top 10 markers.
#### **OPC_1**
**-**: Zmat4 (âˆ’3.66), Zic2 (âˆ’1.01), Zic4 (âˆ’3.17), Zwilch (âˆ’1.45)
**+**: Zfp57 (+1.26), Zkscan16 (+0.83)
immature..?

#### **OPC_2**
**+**: Zgrf1 (+3.26), Zwilch (+2.64), Zfp57 (+0.39), Zfp941 (+0.44)
**-**: Zfp385b (âˆ’4.72), Zic5 (âˆ’3.62)
_Zgrf1, Zwilch_ - proliferation/mitotic 
proliferating OPC?

##### **OPC_3**
**+**: Zmat4 (+3.67), Zic5 (+0.88), Zic2 (+0.77), Zic1 (+1.08), Zic4 (+2.16),  Zkscan16 (+1.30), Zfyve28 (+2.34), Zmym3 (+0.99)
**-**: Zwilch (âˆ’0.77), Zgrf1 (âˆ’1.10)
Zic family transcription factor
differentiation/lineage commitment?

# DEG
```{r [O] DEG, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}

```

